### ISSUE TITLE
HTTP proxy support

### ISSUE THREAD

**Issue by markw83** [rabbitmq/rabbitmq-trust-store#69]
We also face the problem of having a forward proxy between RabbitMQ and our certificate provider. We tried to set the http_proxy, https_proxy and no_proxy settings in the operating system and/or setting this parameters in the rabbitmq-env.conf, but it does not seem to work. We always get the error:

Unable to load certificate list for provider rabbit_trust_store_http_provider, reason: {failed_connect,[{to_address,{"xxx",yyy}},{inet,[inet],etimedout}]}

Is it even possible to configure this for the plugin or RabbitMQ at all?

**Comment by michaelklishin** [comment]
Not sure what "also" means here.

It is possible to configure an HTTP proxy [for the HTTP client](http://erlang.org/doc/man/httpc.html) this project uses. It should be possible to do via RabbitMQ's [advanced.config](http://www.rabbitmq.com/configure.html#configuration-files) but I haven't tried it and the plugin itself
doesn't have any settings it would pass on to the client programmatically.

FWIW this is the first time we see interest in using the HTTP[S] provider used with a proxy, so the state of things is not terribly surprising.

**Comment by markw83** [comment]
Thank you for pointing me to the advanced.config but somehow it does not seem to work. I am no expert in Erlang and thus, do not really know how to set the configuration. I tried the following code in the /etc/rabbitmq/advanced.config (ip is only for testing):

```
[{inets,
  [{httpc,
    [
      {proxy, {{"127.0.0.1", 8080},[]}},
      {https_proxy, {{"127.0.0.1", 8080},[]}}
    ]
  }]
}].
```

I see it added to the generated configuration in /var/lib/rabbitmq/config/generated/rabbitmq.config but I cannot see if it is used. RabbitMQ still works the same as without the configuration. That is, connecting to the remote server works with the above proxy settings which should not be the case (e.g., having this setting in the environment variables, prevents curl from accessing the server).

What am I doing wrong?

**Comment by michaelklishin** [comment]
Your attempt does all the right things.

Looking at the `httpc` source I don't think it's possible to configure the client using the typical application configuration keys. The only way to configure it is to use one of the option setter functions or at request time, so the plugin has to do it.

**Comment by michaelklishin** [comment]
We can't delay `3.7.12` over this but it should be a straightforward thing to add.

**Comment by markw83** [comment]
Thank you for your fast feedback. This means we will wait for the new plugin/RabbitMQ version.

**Comment by markw83** [comment]
Are there any news when this issue will be resolved? We would need to be able to set the proxy information.

**Comment by michaelklishin** [comment]
@hairyhum this seems to be a matter of propagating HTTP[S] proxy settings (endpoint and excludes) to [HTTP options here](https://github.com/rabbitmq/rabbitmq-trust-store/blob/master/src/rabbit_trust_store_http_provider.erl#L64). Am I missing anything?

**Comment by hairyhum** [comment]
@michaelklishin according to [the docs](http://erlang.org/doc/apps/inets/http_client.html#getting-started) it should be set via the [httpc:set_options](http://erlang.org/doc/man/httpc.html#set_options-2) function.

**Comment by markw83** [comment]
I saw `RabbitMQ 3.7.15` is out but it does not include the fix for this issue. Is it moved to `RabbitMQ 3.7.16`? (issue still indicates Milestone 3.7.15)

**Comment by michaelklishin** [comment]
Correct.

**Comment by michaelklishin** [comment]
We have to ship `3.7.16` due to an issue with Shovel reconnection, so rescheduling.

**Comment by michaelklishin** [comment]
Thanks to @dcorbacho this has been merged for `3.7.16`.
